core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -116,7 +117,7 @@ public boolean equals(final Object obj) {
     }
 
     @Override
-    public int hashCode() {
+    public int cachingHashCode() {
         return Objects.hash(getClass(), order, offset, source, iterations, references);
```

Comment:
We may consider doing something here like:
```suggestion
        return Objects.hash(getClass(), order.size, offset, source, iterations.size, references.size);
```
to keep hash calculations simple and quick.

I ran the performance test within ImmutableObjectTest for a `dataBlockCount` of 512 and switched between hash caching on/off and parsestate hash simplified on/off. That resulted into the following performance matrix:
|      | No hash caching  | Hash caching |
| --- | ------------- | ------------- |
| **Original ParseState hash** | undetermined | 730ms |
| **Simplified ParseState hash** | 3600ms | 700ms |

Since the hash caching contributes most to the performance gain, I implemented that first.

Reply:
Since the impact of hash caching itself is so huge, I would wait with further optimization until there is a clear practical case where that needs to be added. That said, I don't oppose this change if you want to add it.